src/compute/src/render.rs

```diff
@@ -241,7 +254,7 @@ pub fn build_compute_dataflow<A: Allocate>(
                         source.storage_metadata.clone(),
                         dataflow.as_of.clone(),
                         snapshot_mode,
-                        dataflow.until.clone(),
+                        until.clone(),
```

Comment:
I'm not sure if this is correct. We are now passing the expiration time as the `until` to the storage source operators. Which means that these operators are not free to produce all times up to the until, and then jump directly to the empty frontier. `expire_stream_at` wouldn't panic in this case because it has an exception for the empty frontier, so we would show invalid data for times >= `until`. I think we still need to pass `dataflow.until` here, or don't ignore the empty frontier in `expire_stream_at`.

Reply:
Yes, the logic needs a rework wrt empty frontiers, as is also obvious from a [CI failure](https://buildkite.com/materialize/test/builds/90934#annotation-01921f5b-c5ae-422c-85cc-777c9092c5cb-error).

Reply:
As we discussed, this is needed in `mfp.evaluate()`.

Reply:
I think Jan has a point, because a source could produce all data at the correct point in time, and then drop its capability. The downstream operator could see the data, plus the frontier advancement at the same time, which would make it hard to reason about what the expiration logic should do.

I think this is mitigated by only allowing the milliseconds timeline, which implies that we won't observe the forward-jumping behavior. But it's very hand-wavy.